### DISCUSSION TITLE
How can I pass custom exceptions?

### DISCUSSION THREAD

**Discussion by peskygnat** [bytedeco/javacpp-presets#1546] [Q&A]
I'm continuing to try and convert a large code base from swig to javacpp.   We have lots of custom C++ exceptions that can be thrown.   I have seamingly failed at figuring out how to handle them in my javacpp preset I'm working on.

here is a sample of the c++ definition of the exception:
```
namespace toplevel{
    namespace component{
        class MyException : public std::exception{
        public :
            inline MyException(std::string& message) : std::exception(message) {}
        };
    }
}
```
I have tried various mappings and even javaText to get the results I wanted
`        infoMap.put(new Info("toplevel::component::MyException").skipDefaults().base("Exception"));
`

but in the end I was hoping to generate something like this
```
        @Namespace("toplevel::component") public static class MyException extends Exception {
            public MyException(@StdString String message) {super(message);}
        }
```

on the other end, if I'm interpreting the jni code correctly, the exception handling is done in the following routine:
```
static JavaCPP_noinline jthrowable JavaCPP_handleException(JNIEnv* env, int i) {
    jstring str = NULL;
    try {
        throw;
    } catch (GENERIC_EXCEPTION_CLASS& e) {
        str = JavaCPP_createStringFromBytes(env, e.GENERIC_EXCEPTION_TOSTRING);
    } catch (...) {
        str = JavaCPP_createStringFromBytes(env, "Unknown exception.");
    }
    jmethodID mid = JavaCPP_getMethodID(env, i, "<init>", "(Ljava/lang/String;)V");
    if (mid == NULL) {
        return NULL;
    }
    return (jthrowable)env->NewObject(JavaCPP_getClass(env, i), mid, str);
}
```

I thought it was appropriate to modify it as follow to pass my exception:

```
static JavaCPP_noinline jthrowable JavaCPP_handleException(JNIEnv* env, int i) {
    jstring str = NULL;
    jclass exception_cls = JavaCPP_getClass(env, i);
    try {
        throw;
    } catch (toplevel::mycomponent::MyException& e) {
        str = JavaCPP_createStringFromBytes(env, e.GENERIC_EXCEPTION_TOSTRING);
        exception_cls = (env)->FindClass("toplevel/component/common/MyException");
    } catch (GENERIC_EXCEPTION_CLASS& e) {
        str = JavaCPP_createStringFromBytes(env, e.GENERIC_EXCEPTION_TOSTRING);
    } catch (...) {
        str = JavaCPP_createStringFromBytes(env, "Unknown exception.");
    }
    jmethodID mid = JavaCPP_getMethodID(env, i, "<init>", "(Ljava/lang/String;)V");
    if (mid == NULL) {
        return NULL;
    }

    return (jthrowable)env->NewObject(exception_cls, mid, str);
}
```

Is my approach to using custom exceptions correct?
Since the java exception is never actually used in the c++ code.  Does it make sense to not have it based off of Pointer?

Is there a way for the presets to generate simple class code not based on Pointer?

Is there a way through comments or depreciated methods exception declaration in the c++ code that is being parsed to add the exception handling in c++?  Similar to my example to JavaCPP_handleException?  
 
Could I extend parser to handle these conditions?

**Comment by saudet** [comment]
The Generator doesn't support custom exceptions like that. It assumes by default it's going to get std::exception and throws a RuntimeException with the string from the what() function, but we can change that by defining the GENERIC_EXCEPTION_CLASS and GENERIC_EXCEPTION_TOSTRING macros:
https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Generator.java#L1118-L1140
To get some other class than RuntimeException, it will also pick up the class from the `throws ...` declaration in Java, but that can't be a Pointer, it needs to be an Exception. That's a limitation of Java, that's not something we can "fix" in JavaCPP.

**Reply by saudet** [reply]
And no unfortunately, the Parser doesn't support generating Exception classes like this, but contributions are welcome!
